### ISSUE TITLE
[FEAT]: Custom OpenAI API Invocation Method

### ISSUE THREAD

**Issue by Kong-Aobo** [sotopia-lab/sotopia#152]
### Description

We are from Alibaba Tongyi Laboratory. Due to certain reasons, we are unable to directly call the OpenAI API and instead need to use our custom method (this situation likely applies to most Chinese research institutions). After reviewing the code, I have identified the parts that need modification as follows.

```python
def obtain_chain(
    model_name: str,
    template: str,
    input_variables: list[str],
    temperature: float = 0.7,
    max_retries: int = 6,
) -> RunnableSerializable[dict[Any, Any], BaseMessage]:
    """
    Using langchain to sample profiles for participants
    """
    model_name = _return_fixed_model_version(model_name)
    ...
    else:
         chat = ChatOpenAI(            # Here!  Maybe we can define a class named CompanyAPIChat 
             model=model_name,      # to replace ChatOpenAI.
             temperature=temperature,
             max_retries=max_retries,
        )

        human_message_prompt = HumanMessagePromptTemplate(
            prompt=PromptTemplate(template=template, input_variables=input_variables)
        )
        chat_prompt_template = ChatPromptTemplate.from_messages([human_message_prompt])
        chain = chat_prompt_template | chat
        return chain
```
After reviewing some of the Langchain documentation, we tried an implementation as follows.
```python
import aiohttp
import requests
from langchain_core.language_models.chat_models import BaseChatModel

async def post_request(url, headers, payload):
    async with aiohttp.ClientSession() as session:
        async with session.post(url, headers=headers, json=payload) as response:
            response_data = await response.json()
    return response_data

class CompanyAPIChat(BaseChatModel):
    model_name: str = Field(default="gpt-4", alias="model")
    temperature: float = Field(default=0.7, type=float)
    max_retries: int = Field(default=5, type=int)
    def __init__(self, model_name: str, temperature: float, max_retries: int):
        super().__init__()
        self.model_name = model_name

    def invoke(self, messages, config, **kwargs):

        print(self.model_name, self.temperature, self.max_retries)
        temp = []
        messages = messages.to_messages()
        for message in messages:
            if isinstance(message, SystemMessage):
                temp.append({"role": "system", "content": message.content})
            elif isinstance(message, AIMessage):
                temp.append({"role": "assistant", "content": message.content})
            else:
                temp.append({"role": "user", "content": message.content})
        messages = temp

        url = ""   # sorry, I can not show it here.
        token = ""
        payload = {
            'model': self.model_name,
            'messages': messages,
            'temperature': self.temperature,
        }
        headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
        response = requests.post(url, headers=headers, json=payload)
        response.raise_for_status()  # Raise an error for bad status codes
        return json.loads(response.text)["data"]["response"]["choices"][0]["message"]["content"]
        
    async def ainvoke(self, messages, config, **kwargs):

        print(self.model_name, self.temperature, self.max_retries)
        temp = []
        messages = messages.to_messages()
        for message in messages:
            if isinstance(message, SystemMessage):
                temp.append({"role": "system", "content": message.content})
            elif isinstance(message, AIMessage):
                temp.append({"role": "assistant", "content": message.content})
            else:
                temp.append({"role": "user", "content": message.content})
        messages = temp

        url = ""    # sorry, I can not show it here.
        token = ""
        payload = {
            'model': self.model_name,
            'messages': messages,
            'temperature': self.temperature,
        }
        headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
        response = await post_request(url, headers=headers, payload=payload)
        print(response)
        # response = await requests.post(url, headers=headers, json=payload)
        # response.raise_for_status()  # Raise an error for bad status codes
        return response["data"]["response"]["choices"][0]["message"]["content"]
         

    @property
    def _llm_type(self) -> str:
        """Get the type of language model used by this chat model. Used for logging purposes only."""
        return "custom"

    def _generate(self):
        pass
```

We design a small unit test and this code passed the test. However, when I replace the original code with our code just like this:
```python
```python
def obtain_chain(
    model_name: str,
    template: str,
    input_variables: list[str],
    temperature: float = 0.7,
    max_retries: int = 6,
) -> RunnableSerializable[dict[Any, Any], BaseMessage]:
    """
    Using langchain to sample profiles for participants
    """
    model_name = _return_fixed_model_version(model_name)
    ...
    else:
        # chat = ChatOpenAI(
        #     model=model_name,
        #     temperature=temperature,
        #     max_retries=max_retries,
        # )
        
        chat = CompanyAPIChat(
            model=model_name,
            temperature=temperature,
            max_retries=max_retries,
        )

        human_message_prompt = HumanMessagePromptTemplate(
            prompt=PromptTemplate(template=template, input_variables=input_variables)
        )
        chat_prompt_template = ChatPromptTemplate.from_messages([human_message_prompt])
        chain = chat_prompt_template | chat
        return chain
```
minimalist_demo.py failed to run. During step-by-step debugging, we found that the code jumps out of the obtain_chain() function when executing the line chat = CompanyAPIChat( model=model_name, temperature=temperature, max_retries=max_retries, ). We suspect there is an issue with the instantiation of this class (perhaps our implementation is not correct in the Sotopia code environment). Since we are not very familiar with Langchain and there are no error messages provided during code execution, we hope to receive your assistance.

Actually, the method of calling API in our lab is so easy:
```python
import requests
import json

token = ''
url = ""
messages = [
    {'role': 'system', 'content': 'You are a helpful assistant.'},
    {'role': 'user', 'content': '你好啊'}
]

payload = {
 'model': 'gpt-4-1106-preview',
 'messages': messages
}

headers = {
    'Authorization': f'Bearer {token}',
    'Content-Type': 'application/json'
}

response = requests.request("POST", url, headers=headers, json=payload)
print(response.text)
```

To sum up, maybe you can help us from these two directions:
1. Could you help us identify the issue with the current implementation? (namely CompanyAPIChat)
2. Given your knowledge of Langchain, is there a simpler way to implement this?

This request is a bit unusual, and I completely understand if you don't want to spend time on it. Regardless, thank you for the effort you've put in.

**Comment by ProKil** [comment]
Hi @Kong-Aobo, thanks for investigating this!

We current support custom APIs that are compatible with openai's chat completion API. And we could also add the support for any publicly available API. Private non-OpenAI-compatible APIs is planned for future versions, but not current priority given the small team we have. 

If your case is the former, can you try to use the new feature introduced in #142? We have tested it with the endpoints hosted with vllm (https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html). Let me know if you need more help with setting it up.

**Comment by ProKil** [comment]
@Kong-Aobo Could you check if the latest PR fits your use case? Feel free to reopen if it doesn't work for you.
